File: DigitalSafe.java
public class DigitalSafe {
    private /*@ spec_public @*/ int safe_state;
    private /*@ spec_public @*/ int attempts;

    //@ public invariant (0 <= safe_state);
    //@ public invariant (safe_state <= 2);
    //@ public invariant (0 <= attempts);
    //@ public invariant (attempts <= 3);
    //@ public invariant ((safe_state == 1) ==> (attempts == 0));
    //@ public invariant ((safe_state == 2) ==> (attempts == 3));

    //@ assignable \nothing;
    //@ ensures safe_state == 0 && attempts == 0;
    public DigitalSafe() {
        this.safe_state = 0;
        this.attempts = 0;
    }

    //@ assignable \nothing;
    //@ ensures \result == safe_state;
    public /*@ pure @*/ int getSafeState() { return safe_state; }

    //@ assignable \nothing;
    //@ ensures \result == attempts;
    public /*@ pure @*/ int getAttempts() { return attempts; }

    //@ requires (safe_state == 0);
    //@ requires (attempts == 0);
    //@ assignable safe_state;
    //@ ensures safe_state == 1;
    public void unlock() {}

    //@ requires (safe_state == 0);
    //@ requires (attempts <= 1);
    //@ assignable attempts;
    //@ ensures attempts == (\old(attempts) + 1);
    public void failAttempt() {}

    //@ requires (safe_state == 0);
    //@ requires (attempts == 2);
    //@ assignable safe_state, attempts;
    //@ ensures safe_state == 2 && attempts == 3;
    public void blockSafe() {}

    //@ requires (safe_state == 1);
    //@ assignable safe_state, attempts;
    //@ ensures safe_state == 0 && attempts == 0;
    public void lock() {}

    //@ requires (safe_state == 2);
    //@ assignable safe_state, attempts;
    //@ ensures safe_state == 0 && attempts == 0;
    public void adminReset() {}
}

File: DigitalSafe.rs
use prusti_contracts::*;

pub struct DigitalSafe {
    pub safe_state: i32,
    pub attempts: i32,
}

impl DigitalSafe {
    #[ensures(result.safe_state == 0 && result.attempts == 0 && (0 <= result.safe_state) && (result.safe_state <= 2) && (0 <= result.attempts) && (result.attempts <= 3) && ((result.safe_state == 1) ==> (result.attempts == 0)) && ((result.safe_state == 2) ==> (result.attempts == 3)))]
    pub fn new() -> Self {
        Self { safe_state: 0, attempts: 0 }
    }

    #[pure]
    #[ensures(result == self.safe_state)]
    pub fn get_safe_state(&self) -> i32 {
        self.safe_state
    }

    #[pure]
    #[ensures(result == self.attempts)]
    pub fn get_attempts(&self) -> i32 {
        self.attempts
    }

    #[requires((0 <= self.safe_state) && (self.safe_state <= 2) && (0 <= self.attempts) && (self.attempts <= 3) && ((self.safe_state == 1) ==> (self.attempts == 0)) && ((self.safe_state == 2) ==> (self.attempts == 3)))]
    #[requires(self.safe_state == 0)]
    #[requires(self.attempts == 0)]
    #[ensures(self.safe_state == 1)]
    #[ensures((0 <= self.safe_state) && (self.safe_state <= 2) && (0 <= self.attempts) && (self.attempts <= 3) && ((self.safe_state == 1) ==> (self.attempts == 0)) && ((self.safe_state == 2) ==> (self.attempts == 3)))]
    pub fn unlock(&mut self) {
        self.safe_state = 1;
    }

    #[requires((0 <= self.safe_state) && (self.safe_state <= 2) && (0 <= self.attempts) && (self.attempts <= 3) && ((self.safe_state == 1) ==> (self.attempts == 0)) && ((self.safe_state == 2) ==> (self.attempts == 3)))]
    #[requires(self.safe_state == 0)]
    #[requires(self.attempts < 2)]
    #[ensures(self.attempts == old(self.attempts) + 1)]
    #[ensures((0 <= self.safe_state) && (self.safe_state <= 2) && (0 <= self.attempts) && (self.attempts <= 3) && ((self.safe_state == 1) ==> (self.attempts == 0)) && ((self.safe_state == 2) ==> (self.attempts == 3)))]
    pub fn fail_attempt(&mut self) {
        let pre_attempts = self.attempts;
        self.attempts = pre_attempts + 1;
    }

    #[requires((0 <= self.safe_state) && (self.safe_state <= 2) && (0 <= self.attempts) && (self.attempts <= 3) && ((self.safe_state == 1) ==> (self.attempts == 0)) && ((self.safe_state == 2) ==> (self.attempts == 3)))]
    #[requires(self.safe_state == 0)]
    #[requires(self.attempts == 2)]
    #[ensures(self.safe_state == 2 && self.attempts == 3)]
    #[ensures((0 <= self.safe_state) && (self.safe_state <= 2) && (0 <= self.attempts) && (self.attempts <= 3) && ((self.safe_state == 1) ==> (self.attempts == 0)) && ((self.safe_state == 2) ==> (self.attempts == 3)))]
    pub fn block_safe(&mut self) {
        self.safe_state = 2;
        self.attempts = 3;
    }

    #[requires((0 <= self.safe_state) && (self.safe_state <= 2) && (0 <= self.attempts) && (self.attempts <= 3) && ((self.safe_state == 1) ==> (self.attempts == 0)) && ((self.safe_state == 2) ==> (self.attempts == 3)))]
    #[requires(self.safe_state == 1)]
    #[ensures(self.safe_state == 0 && self.attempts == 0)]
    #[ensures((0 <= self.safe_state) && (self.safe_state <= 2) && (0 <= self.attempts) && (self.attempts <= 3) && ((self.safe_state == 1) ==> (self.attempts == 0)) && ((self.safe_state == 2) ==> (self.attempts == 3)))]
    pub fn lock(&mut self) {
        self.safe_state = 0;
        self.attempts = 0;
    }

    #[requires((0 <= self.safe_state) && (self.safe_state <= 2) && (0 <= self.attempts) && (self.attempts <= 3) && ((self.safe_state == 1) ==> (self.attempts == 0)) && ((self.safe_state == 2) ==> (self.attempts == 3)))]
    #[requires(self.safe_state == 2)]
    #[ensures(self.safe_state == 0 && self.attempts == 0)]
    #[ensures((0 <= self.safe_state) && (self.safe_state <= 2) && (0 <= self.attempts) && (self.attempts <= 3) && ((self.safe_state == 1) ==> (self.attempts == 0)) && ((self.safe_state == 2) ==> (self.attempts == 3)))]
    pub fn admin_reset(&mut self) {
        self.safe_state = 0;
        self.attempts = 0;
    }
}
